### ISSUE TITLE
Ordering guarantees with multiple streaming results

### ISSUE THREAD

**Issue by Tragetaschen** [aspnet/SignalR#1321]
In `alpha2`, I have three streaming methods (`IObservable`) in a Hub driven by System.Reactive's `Observable.FromEvent` and the latter converts three different events A, B, C.
In another part of the program, I run a single thread that invokes those three events in order A, B, C. When I tap (`Do`) into the Observables, I can see the events in the correct order.

On the client side using two Chrome instances, I observe the frames of the underlying web socket connection. Most of the time, I get frames corresponding to the event invocations in order A, B, C, but sometimes the order changes to B, C, A. This change is on a per-instance basis. One instance sees this, the other one gets the correct order A, B, C.

Before I explicitly set the `IScheduler` for the observables to `ImmediateScheduler.Instance`, this problem was much worse. I'm assuming that the Hub methods are called within the ThreadPool and System.Reactive chooses its ThreadPool based scheduler, thus making the order mostly random. Now, my observable should be entirely synchronous.

Are there (going to be) any ordering guarantees for the above scenario in RTM SignalR Core? For now, I'll merge my three observables into one doing explicit dispatching on the other end. This seems to keep the expected order.

**Comment by BrennanConroy** [comment]
Whatever order the observable returned from your hub method returns values is the order they get placed on the wire. So it should be guaranteed ordering.

If you could provide an example I can take a look.

**Comment by Tragetaschen** [comment]
Yes, for each Observable on its own, that is what I'm seeing.

The question is for one hub with three observables A,B,C.
If they fire in that order, does SignalR guarantee that order on the wire? What I'm seeing is that most of the time yes, but sometimes they arrive at the client in a different order.

**Comment by BrennanConroy** [comment]
One hub with three observables? Could you share how you're doing that?

**Comment by Tragetaschen** [comment]
Huh?

```cs
class MyHub : Hub
{
    public IObservable<int> A();
    public IObservable<int> B();
    public IObservable<int> C();
}
```

**Comment by anurse** [comment]
No, there is no ordering guarantee here. If you invoke `A` and `B` simultaneously, you will receive the results interleaved. This is the designed behavior, after all if `A` emits items indefinitely until cancelled, or has long durations between items, we wouldn't want to block the processing of other invocations from the client. If you need ordering, you should attach ordering data to the items you are emitting and reorder on the client.

**Comment by davidfowl** [comment]
That's by design. Each stream is independent. The SignalR protocol multiplexes results over the connection as soon as they are available.

**Comment by Tragetaschen** [comment]
So in layman's terms: There's one queue per connection/stream pair and those queues flush independently onto the actual connection guaranteeing correct order per stream, but don't give anything beyond that.

**Comment by davidfowl** [comment]
Correct. There's no ordering across method calls.

**Comment by Tragetaschen** [comment]
That's fine. Then my merging of observables is the correct solution with the right guarantees.
